framework/core/allocated.cpp

```diff
@@ -0,0 +1,235 @@
+/* Copyright (c) 2021-2024, NVIDIA CORPORATION. All rights reserved.
+ * Copyright (c) 2024, Bradley Austin Davis. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "allocated.h"
+
+namespace vkb
+{
+
+namespace allocated
+{
+
+VmaAllocator &get_memory_allocator()
+{
+	static VmaAllocator memory_allocator = VK_NULL_HANDLE;
+	return memory_allocator;
+}
+
+void init(const VmaAllocatorCreateInfo &create_info)
+{
+	VkResult result = vmaCreateAllocator(&create_info, &get_memory_allocator());
+	if (result != VK_SUCCESS)
+	{
+		throw VulkanException{result, "Cannot create allocator"};
+	}
+}
+
+void shutdown()
+{
+	auto &allocator = get_memory_allocator();
+	if (allocator != VK_NULL_HANDLE)
+	{
+		VmaTotalStatistics stats;
+		vmaCalculateStatistics(allocator, &stats);
+		LOGI("Total device memory leaked: {} bytes.", stats.total.statistics.allocationBytes);
+		vmaDestroyAllocator(allocator);
+		allocator = VK_NULL_HANDLE;
+	}
+}
+
+AllocatedBase::AllocatedBase(const VmaAllocationCreateInfo &alloc_create_info) :
+    alloc_create_info(alloc_create_info)
+{
+}
+
+AllocatedBase::AllocatedBase(AllocatedBase &&other) noexcept :
+    alloc_create_info(std::exchange(other.alloc_create_info, {})),
+    allocation(std::exchange(other.allocation, {})),
+    mapped_data(std::exchange(other.mapped_data, {})),
+    coherent(std::exchange(other.coherent, {})),
+    persistent(std::exchange(other.persistent, {}))
+{
+	assert(other.allocation == VK_NULL_HANDLE);
+	assert(other.mapped_data == nullptr);
```

Comment:
Yeah, I don't know why I did that.  It doesn't seem to be sourced from any of the code I ported.  It's possible I was migrating code that didn't use `std::exchange`, like the old image or buffer move constructor and added these asserts and then failed to remove them when I converted to use of `std::exchange`.  

I don't want to fix it right now though because I don't want to invalidate the approvals for something that doesn't have any negative impact.  If anything comes along that does force me to update the commits, I'll try to get this fixed too.

Reply:
It's okay to resolve. Just ping us and we'll approve again.

Reply:
fixed